### PR TITLE
(Revised) show prefer_dedicated_media_query_method lint for indirect propery access too

### PR DESCRIPTION
**Pull request by imsamgarg**
# Description

The current `prefer_dedicated_media_query_method` lint only works for direct MediaQuery.of(context).property access. For cases like these, it does not work.

```dart
final mq = MediaQuery.of(context);
final size = mq.size;
```

So I have implemented the solution using the context.registry.addPrefixedIdentifier() for both link and fix.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactor
- [ ] Documentation
- [ ] Other

## Checklist

- [x] I have read the [CONTRIBUTING.md][contributing_link] document.
- [x] I have performed a self-review of my code.
- [ ] I have linked the issue ticket in the description.
- [ ] I have made the necessary changes to the documentation.
- [x] I have checked the formatting with `dart format .`
- [x] I have analyzed my code with `dart analyze .`
- [x] I have run `dart run custom_lint example` to check for any custom linting issues.

[contributing_link]: https://github.com/charlescyt/pyramid_lint/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by imsamgarg** [comment]
I have added the fix for this case, please review it.

void fn(MediaQueryData mq){
  final size = mq.size; // lint should not warn here
}

**Comment by charlescyt** [comment]
Hello! Thanks for the update. I was trying to fix the quick fix such that it works when the `BuildContext` variable name is not `context`.

Here's what I got. First I have a method to get the `AstNode` from a `Element`, adapted from [stackoverflow](https://stackoverflow.com/questions/57030992/how-do-i-get-the-astnode-from-some-element-in-dart-analyzer-source-gen).

```dart
AstNode? getAstNodeFromElement(Element element) {
  final session = element.session;
  if (session == null) return null;

  final elementLibrary = element.library;
  if (elementLibrary == null) return null;

  final parsedLibraryResult =
      session.getParsedLibraryByElement(elementLibrary) as ParsedLibraryResult;
  final elementDeclarationResult =
      parsedLibraryResult.getElementDeclaration(element);

  return elementDeclarationResult?.node;
}
```

Then in the quick fix method, we try to get the `VariableDeclaration` and check if the variable is initialised with `MediaQuery.of(...)`, and also extract the BuildContext's name and use it in the quick fix. 

```dart
final methodName = node.identifier.name;

final prefixElement = node.prefix.staticElement;
if (prefixElement is! LocalVariableElement) return;

final prefixNode = getAstNodeFromElement(prefixElement);
if (prefixNode is! VariableDeclaration) return;

// Check if the initializer is MediaQuery.of(...)
final initializer = prefixNode.initializer;
if (initializer is! MethodInvocation) return;
if (initializer.methodName.name != 'of') return;

final initializerType = initializer.realTarget?.staticType;
if (initializerType != null &&
    mediaQueryChecker.isExactlyType(initializerType)) return;

if (initializer.argumentList.arguments.length != 1) return;

final argument = initializer.argumentList.arguments.first;
if (argument is! SimpleIdentifier) return;

final argumentName = argument.name;
final dedicatedMethod = 'MediaQuery.${methodName}Of($argumentName)';
```

It works but I am not sure if this is the best way to implement it. If you have any ideas, I would be glad to know.

File: lib/src/lints/flutter/prefer_dedicated_media_query_method.dart [hunk 1]
```diff
@@ -54,6 +54,38 @@ class PreferDedicatedMediaQueryMethod extends DartLintRule {
...
+        packageName: 'flutter',
+      );
+
+      if (targetType == null ||
+          !mediaQueryDataChecker.isExactlyType(targetType)) return;
+
+      /// Lint should only work if the mediaQueryData variable is declared locally.
+      if (node.prefix.staticElement! is! LocalVariableElement) {
+        return;
+      }
```
Comment (charlescyt) [inline lib/src/lints/flutter/prefer_dedicated_media_query_method.dart:71]:
```suggestion
      if (node.prefix.staticElement is! LocalVariableElement) {
        return;
      }
```

File: lib/src/lints/flutter/prefer_dedicated_media_query_method.dart [hunk 2]
```diff
@@ -54,6 +54,38 @@ class PreferDedicatedMediaQueryMethod extends DartLintRule {
   ) {
     if (!context.pubspec.isFlutterProject) return;
 
+    context.registry.addPrefixedIdentifier((node) {
+      final targetType = node.prefix.staticType;
+
+      const mediaQueryDataChecker = TypeChecker.fromName(
+        'MediaQueryData',
+        packageName: 'flutter',
+      );
```
Comment (charlescyt) [inline lib/src/lints/flutter/prefer_dedicated_media_query_method.dart:63]: We should probably move this to the type checker file

Comment (imsamgarg) [reply lib/src/lints/flutter/prefer_dedicated_media_query_method.dart]: Done

**Comment by imsamgarg** [comment]
> Hello! Thanks for the update. I was trying to fix the quick fix such that it works when the `BuildContext` variable name is not `context`.
> 
> Here's what I got. First I have a method to get the `AstNode` from a `Element`, adapted from [stackoverflow](https://stackoverflow.com/questions/57030992/how-do-i-get-the-astnode-from-some-element-in-dart-analyzer-source-gen).
> 
> ```dart
> AstNode? getAstNodeFromElement(Element element) {
>   final session = element.session;
>   if (session == null) return null;
> 
>   final elementLibrary = element.library;
>   if (elementLibrary == null) return null;
> 
>   final parsedLibraryResult =
>       session.getParsedLibraryByElement(elementLibrary) as ParsedLibraryResult;
>   final elementDeclarationResult =
>       parsedLibraryResult.getElementDeclaration(element);
> 
>   return elementDeclarationResult?.node;
> }
> ```
> 
> Then in the quick fix method, we try to get the `VariableDeclaration` and check if the variable is initialised with `MediaQuery.of(...)`, and also extract the BuildContext's name and use it in the quick fix.
> 
> ```dart
> final methodName = node.identifier.name;
> 
> final prefixElement = node.prefix.staticElement;
> if (prefixElement is! LocalVariableElement) return;
> 
> final prefixNode = getAstNodeFromElement(prefixElement);
> if (prefixNode is! VariableDeclaration) return;
> 
> // Check if the initializer is MediaQuery.of(...)
> final initializer = prefixNode.initializer;
> if (initializer is! MethodInvocation) return;
> if (initializer.methodName.name != 'of') return;
> 
> final initializerType = initializer.realTarget?.staticType;
> if (initializerType != null &&
>     mediaQueryChecker.isExactlyType(initializerType)) return;
> 
> if (initializer.argumentList.arguments.length != 1) return;
> 
> final argument = initializer.argumentList.arguments.first;
> if (argument is! SimpleIdentifier) return;
> 
> final argumentName = argument.name;
> final dedicatedMethod = 'MediaQuery.${methodName}Of($argumentName)';
> ```
> 
> It works but I am not sure if this is the best way to implement it. If you have any ideas, I would be glad to know.

It works perfectly fine. It also handles this case as well.

```dart
 Scaffold(
      body: Builder(
        builder: (BuildContext context) {
          return Container(
            // expect_lint: prefer_dedicated_media_query_method
            height: mq.size.height, // mq declared with .of(c)
          );
        },
      ),
    );
```

I have implemented a solution that involves searching for the declaration in the enclosing block recursively. 

  ```dart
  // Main logic
    if (block != null) {
        final mqVarDeclaration = block.statements
            // Need to check only the preceding statements.
            .where((element) => element.offset < node.offset)
            .whereType<VariableDeclarationStatement>()
            .map((e) => e.variables.variables)
            .flattened
            .firstWhereOrNull(
              (e) =>
                  e.name.lexeme == node.prefix.name &&
                  mediaQueryDataChecker.isExactlyType(e.declaredElement!.type),
            );

        final contextArgument =
            (mqVarDeclaration?.initializer as MethodInvocation?)
                ?.argumentList
                .arguments
                .whereType<SimpleIdentifier>()
                .firstWhereOrNull(
                  (e) =>
                      e.staticType != null &&
                      buildContextTypeChecker.isExactlyType(e.staticType!),
                );

        contextArgName = contextArgument?.name ?? contextArgName;
      }
```

But I think it's quite verbose and error-prone. the getAstNodeFromElement solution is more concise I think. We can go with that.

**Pull request merged**